Unify keyword argument names for convolution and pooling

ebenolson opened the pull request:
This is to close #207 

I've made changes to use `filter_size`, `pool_size` and `stride` throughout all conv and pooling layers. 

Added a helper function `as_tuple` in `utils.py` which takes an input and a length and returns a tuple of that length, and use it in the 2D layers to accept either a number or tuple for `pool_size` and `stride`. 

I also added some tests for the cuda_convnet and dnn pooling layers, and updated the examples to use `pool_size`.

f0k commented on lasagne/layers/conv.py:
```diff
@@ -133,8 +134,8 @@ def __init__(self, incoming, num_filters, filter_size, strides=(1, 1),
-        self.filter_size = filter_size
-        self.strides = strides
+        self.filter_size = as_tuple(filter_size, 2)
+        self.stride = as_tuple(stride, 2)
```
Nice. That looks very clean.

f0k commented on lasagne/layers/cuda_convnet.py:
```diff
@@ -198,19 +202,33 @@ def get_output_for(self, input, **kwargs):
 
+        if 'pad' in kwargs:
+            raise NotImplementedError("MaxPool2DCCLayer does not "
+                                      "support padding")
```
Good idea, that's more descriptive than letting the base class constructor raise something about an unknown kwarg.

ebenolson commented:
just realized my tests are not testing very much :/ 

need to think a bit more about how to do this cleanly.

benanne commented on lasagne/layers/conv.py:
```diff
@@ -48,7 +49,7 @@ def __init__(self, incoming, num_filters, filter_length, stride=1,
 
         self.num_filters = num_filters
-        self.filter_length = filter_length
+        self.filter_size = filter_size
```
For consistency with the 2D (and possibly in the future 3D) case, I think it would be nice to do `as_tuple(filter_size, 1)` here, so that the `filter_size` argument can also accept single element tuples. This will also require some other modifications to the code, as it is currently assumed to be an integer (and it would become a single element tuple).

ebenolson replied:
If we're going to use tuples everywhere, what do you think about standardizing on `filter_shape` and `pool_shape` instead? Seems like it would be more consistent with `input_shape` and `output_shape`.

benanne replied:
That's a very good point, although it makes it slightly more awkward when specifying it as an integer. `pool_shape=3` looks a bit weird to me because I've been conditioned to expect a shape to look like a tuple, and I think this goes for most numpy / Theano users :)

But of course there is a lot to be said for consistency. Would love to hear @f0k's thoughts on this (and well, anyone else's)!

f0k replied:
Wohoo, another name discussion! ;) `filter_shape` would also be consistent with Theano, but we don't necessarily need that. As a counterexample, `scipy.ndimage.median_filter` has a `size` keyword that can be a scalar or tuple. And I agree that `pool_shape=3` looks weird, as if we would enumerate possible shapes (1: square, 2: triangle, 3: circle).

We could just make it a rule that things that *must* be tuples are `shape`, and things that can also be scalars are `size`. If we follow that everywhere, this would still count as consistent.

Regarding the original comment, I thought about that, too, but I think `as_tuple(filter_size, 1)` would make the rest of the code a bit harder to read. It could be helpful for refactoring the code, though (there's a lot of duplication between the convolution classes).

benanne replied:
I think if we're going to accept integers and tuples for the 2D and 3D case, there is no reason to only accept integers for the 1D case. It seems more consistent to use `as_tuple` everywhere imo.

> We could just make it a rule that things that must be tuples are `shape`, and things that can also be scalars are `size`. If we follow that everywhere, this would still count as consistent.

Works for me. But using `shape` everywhere is also fine really. It's not that crucial.

benanne commented:
Looks great so far :) Thanks a lot! I will have a more detailed look at the code when you're happy with the tests.

benanne commented:
Looks like there are 3 things left to do on this one
- [ ] use as_tuple in the 1D case as well
- [ ] (optionally) rename the arguments from `_size` to `_shape` (do whatever you prefer here, I'm fine either way)
- [ ] fix tests

If there's any of these I can take off your hands let me know. I believe it should be possible to send PRs for PRs ;)

ebenolson commented:
ok, I'll try to get this wrapped up later today.

ebenolson commented:
Reworked `test_pool.py`, due to all the restrictions on `cuda_convnet` it seemed simpler to make separate tests for each implementation. I also added some tests for `Conv1DLayer`, at some point I want to clean up the conv tests more but for now it's better than nothing.

Changed the 1D conv and pool layers to use `as_tuple` as well.

I added the `pad` argument to `MaxPool2DDNNLayer` since it seems to work now (I guess v2 wasn't out when that was written?)

I decided to stick with `_size`, I think @f0k's suggestion to standardize on `_size` for things that may be either tuples or scalar makes sense.

f0k commented on lasagne/layers/cuda_convnet.py:
```diff
@@ -198,34 +202,51 @@ def get_output_for(self, input, **kwargs):
-            raise RuntimeError("MaxPool2DCCLayer does not support "
-                               "ignore_border.")
+            raise NotImplementedError("MaxPool2DCCLayer does not support "
+                                      "ignore_border.")
```
Makes sense. It's even something that one *could* add here if anybody ever needs it.

f0k commented on lasagne/layers/pool.py:
```diff
@@ -236,38 +224,45 @@ class FeaturePoolLayer(Layer):
-        axis: the axis along which to pool. The default value of 1 works
-        for DenseLayer and Conv*DLayers
-        pool_function: the pooling function to use
+        Instrideantiates the layer.
```
Haha, nice neologism! Freudian slip? :D

f0k commented on lasagne/layers/pool.py:
```diff
@@ -145,11 +151,11 @@ def __init__(self, incoming, ds, st=None,
             The layer feeding into this layer, or the expected input shape.
-        ds : integer or iterable
+        pool_size : integer or iterable
             The length of the pooling region in each dimension
```
That's quite elegant, actually. The text fits both the integer and iterable case.

f0k commented on lasagne/utils.py:
```diff
@@ -66,6 +66,28 @@ def unique(l):
+    -----------
+    x : value or iterable
+    N : integer
+        length of the desired tuple
```
```
Returns:
--------
tuple
    ``tuple(x)`` if `x` is iterable, ``(x,) * N`` otherwise.
```

f0k commented:
Thanks, looks very clean and mergeable to me, except for two minor comments!

ebenolson commented:
oops, thought I'd caught all those. thanks for reviewing.

benanne commented:
Looking good! Will merge tomorrow then :)

benanne commented:
Merging! I'll make a post on the mailing list about the changed interface.
